### PR TITLE
[login] module upgraded for react

### LINKED ISSUE #5214: [login] module rewrite with react
Redmine: https://redmine.cbrain.mcgill.ca/issues/16050

**Description:**
Make the login form use react instead of smarty.

#4395

### PR DESCRIPTION
**Pull request by maltheism**
## Brief summary of changes

Upgraded module to use react. I had to fix some things in LORIS to get this all to work and there are multiple changes in some JSX files as well.

What was updated to react:
Login, password expired, request account, and reset password.

Summary:
Turns out the Content-Security policy was broken for google captcha and I made a fix to include the iframe that recaptcha uses. Major was also broken inside the php/libraries/Email.class.inc and I made some fixes. That had to do with the password reset.

Note:
This PR includes the bugfix PR #4386 and so this PR can function.

#### Testing instructions (if applicable)

1. Checkout PR.
2. Login
3. Logout
4. Request Account
5. Recover Password

#### Links to related tickets (GitHub, Redmine, ...)

* Redmine #16050
* Resolves #5828 
* Resolves #5214

### CODE REVIEW

**Comment by johnsaigle** [comment]
> 1) LorisLoginTest::testLoginFailure
NoSuchElementException: Unable to locate element: *[name='username']

I've commented this before on your closed PRs, but I'll say it again.

This issue is caused by the test in Travis trying to find an element with the HTML attribute "name" equal to "username". If this value was present in smarty but is not present in your new React file, then Travis fails because it's unable to locate the HTML elements necessary to complete its test.

To fix this error (as well as the issue sin your other React PRs), you need to:

1. Update the integration test file in the module to have the name(s) equal to the new HTML element(s) you created, OR
2. Change the new HTML elements in your React code to match the tests.

Basically, these values have to be the same.

**Comment by maltheism** [comment]
@johnsaigle "[Basically](https://dev.to/meeshkan/how-to-remove-condescending-language-from-documentation-4a5p)" on a side note this PR used to be failing because of some reason related to the browser version our test suite was running is what @kongtiaowang told me. The other react PRs were passing such as Dashboard before the recent release. I sort of know how to make the tests pass but typically I have trouble with the test code because it seems to sometimes ignore the selectors I supply.

**Comment by johnsaigle** [comment]
"In summary" would've been a better word choice.

It would be helpful to include some discussion points from previous PRs on your new PRs. Choosing to close PRs instead of rebasing means that context from previous reviews get lost which can cause some miscommunication and double work. 

The reason I'm commenting on Travis errors is because you added the label that is asking for someone to explain Travis. If the problem is caused by an issue with the test suite then I think it would be more helpful, as a reviewer, for you to add `Discussion Required`, `Blocked`, etc. as appropriate.

Could you link to the previous discussion and/or PRs that address the problem? If there's a problem with Travis/Selenium/React, could you open an issue and link to it?

**Comment by maltheism** [comment]
@johnsaigle I was just referencing the condescending word in jest because you brought it to the attention at the Loris meeting. I frequently use "basically" and I always now remember I shouldn't because if it's condescending in documentation it kind of the same elsewhere. 😅

The Travis/Selenium/React discussion I had with @kongtiaowang happened offline and he said that one of his PRs was updating Selenium to fix the problem. I'll ask @kongtiaowang when he gets back about what happened but I think you're right that most travis errors in these react PRs can be fixed by adding the name attribute. I've been busy in the meantime and I thought to add the travis tag for my interpretation tests need to be fixed.

**Comment by johnsaigle** [comment]
I think Travis is indicating that it can't find the elements, not that they're broken. It makes sense that they appear to work manually because it's responding to your clicks, even if the names have changed. The integration tests won't know that though. 

I think for the sake of clarity it would be worth opening an issue with a summary of your conversation with Shen and marking your PRs with Blocked if you're confident the issue is caused by Travis/Selenium. Or you can use Needs Work to indicate that basically you don't need anyone to review your PRs yet.

**Comment by maltheism** [comment]
Hi @johnsaigle I just had some time to look at this PR and the name used to find the element does actually exist in the react component. So I'm not really sure right now. @kongtiaowang I would be really grateful for help with solving these test cases.

**Comment by kongtiaowang** [comment]
@maltheism This PR should wait to this PR https://github.com/aces/Loris/pull/5679 to pass the react test.

**Comment by johnsaigle** [comment]
Adding Blocked based on Shen's comment

**Comment by kongtiaowang** [comment]
@maltheism Could u fix this confilct? I will rewrite thsi test case later.

**Comment by maltheism** [comment]
@kongtiaowang fixed. 🙂

**Comment by kongtiaowang** [comment]
@maltheism you have changed the component that causes side effects of conflict_resolver.
Screen Shot 2020-01-31 at 11 52 38 AM

**Comment by maltheism** [comment]
Thanks @kongtiaowang, I fixed it by changing the logic. That change is intended for modifying thee css of the login module.

**Comment by johnsaigle** [comment]
@kongtiaowang Can you take a look at these integration test errors? I am testing them locally and everything seems to be working.

File: modules/login/php/endpoints/authentication.class.inc [hunk 1]
```diff
@@ -0,0 +1,453 @@
...
+/**
+ * The NDB_Page for the POST requests.
+ *
+ * @category Loris
+ * @package  Login
+ * @author   Alizée Wickenheiser <alizee.wickenheiser@mcin.ca>
+ * @license  http://www.gnu.org/licenses/gpl-3.0.txt GPLv3
+ * @link     https://www.github.com/aces/Loris/
+ */
+class Authentication extends Endpoint implements ETagCalculator
```
Comment (johnsaigle) [inline modules/login/php/endpoints/authentication.class.inc]: If this class is going to use endpoints, I think it makes sense to split them into different endpoints for password reset, account request, and normal logins. Doing a password reset is different from authenticating.

@xlecours maybe you could comment on this? I'm not fully confident about what makes the most sense for endpoints and the API.

File: modules/login/php/endpoints/authentication.class.inc [hunk 2]
```diff
@@ -0,0 +1,453 @@
...
+
+use \Psr\Http\Message\ServerRequestInterface;
+use \Psr\Http\Message\ResponseInterface;
+use \LORIS\Middleware\ETagCalculator;
+use \LORIS\api\Endpoint;
+use SinglePointLogin;
+use State;
+
+/**
+ * The NDB_Page for the POST requests.
```
Comment (johnsaigle) [inline modules/login/php/endpoints/authentication.class.inc]: The function comment need to be updated/cleaned up to match the functionality of this file.

File: modules/login/php/endpoints/authentication.class.inc [hunk 3]
```diff
@@ -0,0 +1,453 @@
+<?php declare(strict_types=1);
+/**
```
Comment (johnsaigle) [inline modules/login/php/endpoints/authentication.class.inc]: You can delete this comment block if you want

File: modules/login/php/endpoints/authentication.class.inc [hunk 4]
```diff
@@ -0,0 +1,453 @@
...
+
+    /**
+     * Initialize setup, the extra values for the
+     * create timepoint form. (psc & errors)
+     *
+     * @param ServerRequestInterface $request The incoming PSR7 request.
+     *
+     * @return ResponseInterface The outgoing PSR7 response
+     */
+    private function _handleGET(ServerRequestInterface $request) : ResponseInterface
```
Comment (johnsaigle) [inline modules/login/php/endpoints/authentication.class.inc]: Each of password reset, request account, and login make more sense as POST requests than GET requests.

File: modules/login/php/endpoints/authentication.class.inc [hunk 5]
```diff
@@ -0,0 +1,453 @@
...
+    {
+        // Login Setup.
+        $loginData = array();
+        $login     = $_SESSION['State']->getProperty('login');
+
+        // Previous login failed.
+        if (!empty($login->_lastError)) {
+            $loginData['error'] = $login->_lastError;
+        }
+        $config      =& \NDB_Config::singleton();
```
Comment (johnsaigle) [inline modules/login/php/endpoints/authentication.class.inc]: Should use Factory instead

File: modules/login/php/endpoints/authentication.class.inc [hunk 6]
```diff
@@ -0,0 +1,453 @@
...
+     */
+    function resetPassword(array $values) : array
+    {
+        // Always claim success to avoid providing user info to attackers.
+        $values['success'] = 'Password reset. '
+            . 'You should receive an email within a few minutes.';
+
+        // Create the user object.
+        $username = $values['username'];
+        $user     =& \User::singleton($username);
```
Comment (johnsaigle) [inline modules/login/php/endpoints/authentication.class.inc]: Factory

File: modules/login/php/endpoints/authentication.class.inc [hunk 7]
```diff
@@ -0,0 +1,453 @@
...
+            $email = $user->getData('Email');
+            if (empty($email)) {
+                $values['error'] = 'Not a valid user';
+                return $values;
+            }
+
+            try {
+                // Check if valid email.
+                if ($user->isEmailValid()) {
+                    $config = \NDB_Config::singleton();
```
Comment (johnsaigle) [inline modules/login/php/endpoints/authentication.class.inc]: Factory

File: modules/login/php/endpoints/authentication.class.inc [hunk 8]
```diff
@@ -0,0 +1,453 @@
...
+    /**
+     * Process request account form.
+     *
+     * @param array $values the user data received.
+     *
+     * @return array
+     */
+    function requestAccount(array $values) : array
+    {
+        $config           =& \NDB_Config::singleton();
```
Comment (johnsaigle) [inline modules/login/php/endpoints/authentication.class.inc]: Factory

File: modules/login/php/endpoints/authentication.class.inc [hunk 9]
```diff
@@ -0,0 +1,453 @@
...
+    }
+
+    /**
+     * Process request account form.
+     *
+     * @param array $values the user data received.
+     *
+     * @return array
+     */
+    function requestAccount(array $values) : array
```
Comment (johnsaigle) [inline modules/login/php/endpoints/authentication.class.inc]: In general these functions would probably be better written using PSR7 RequestInterfaces and extracting the parameters from the `$request` variable. I feel like the `$values` array processing is leftover from smarty templates.

File: modules/login/php/endpoints/authentication.class.inc [hunk 10]
```diff
@@ -0,0 +1,453 @@
...
+    {
+        $config           =& \NDB_Config::singleton();
+        $reCAPTCHAPrivate = $config->getSetting('reCAPTCHAPrivate');
+
+        // Verify reCAPTCHA
+        if (isset($values['g-recaptcha-response']) && isset($reCAPTCHAPrivate)) {
```
Comment (johnsaigle) [inline modules/login/php/endpoints/authentication.class.inc]: An error should probably thrown if these values aren't set in the request. Right now you could avoid this CAPTCHA check by submitting a request manually because the back-end doesn't do any validation.

File: modules/login/php/endpoints/authentication.class.inc [hunk 11]
```diff
@@ -0,0 +1,453 @@
...
+                $values['g-recaptcha-response'],
+                $_SERVER['REMOTE_ADDR']
+            );
+            if (!$resp->isSuccess()) {
+                $values['error'] = 'Please complete the reCaptcha!';
+                return $values;
+            }
+        }
+
+        $DB       = \Database::singleton();
```
Comment (johnsaigle) [inline modules/login/php/endpoints/authentication.class.inc]: Factory

File: modules/login/php/endpoints/authentication.class.inc [hunk 12]
```diff
@@ -0,0 +1,453 @@
...
+            $plaintext = htmlspecialchars_decode($values['password']);
+            // Check if new password is same as old password.
+            if (!$user->isPasswordDifferent($plaintext)) {
+                $values['error'] = 'You cannot keep the same password';
+                return $values;
+            }
+            // Update password
+            $user->updatePassword(
+                new \Password(
+                    htmlspecialchars_decode($values['password'])
```
Comment (johnsaigle) [inline modules/login/php/endpoints/authentication.class.inc]:
```suggestion
                    $plaintext
```

This is already decoded above.

File: modules/login/php/endpoints/authentication.class.inc [hunk 13]
```diff
@@ -0,0 +1,453 @@
...
+     * an incoming HTTP request.
+     *
+     * @param ServerRequestInterface $request The incoming PSR7 request.
+     *
+     * @return string The value to use for the ETag header.
+     */
+    public function ETag(ServerRequestInterface $request): string
+    {
+        if ($request->getMethod() === 'POST') {
+            return 'Etag not calculated on POST request';
```
Comment (johnsaigle) [inline modules/login/php/endpoints/authentication.class.inc]: Should return `''` I think based on Dave's review in another PR

File: modules/login/php/login.class.inc [hunk 14]
```diff
@@ -3,66 +3,47 @@
  * Implements the main LORIS user login page to handle authentication
  * of users.
  *
- * PHP version 5
+ * PHP version 7
```
Comment (johnsaigle) [inline modules/login/php/login.class.inc]: This comment block can be deleted if you want

File: modules/login/php/module.class.inc [hunk 15]
```diff
@@ -3,35 +3,67 @@
  * This serves as a hint to LORIS that this module is a real module.
  * It does nothing but implement the module class in the module's namespace.
  *
- * PHP Version 5
+ * PHP Version 7
```
Comment (johnsaigle) [inline modules/login/php/module.class.inc]: Can delete this comment block

File: modules/login/php/module.class.inc [hunk 16]
```diff
@@ -3,35 +3,67 @@
...
 class Module extends \Module
 {
+    /**
+     * The default handler for a module routes to a page of the first component of
+     * the URL's name, or the same name as the module if serving the root of the
+     * module.
+     *
+     * Modules extend PrefixRouter, so that the parent handler knows that the
+     * css, static, and js routes should be routed through the file system to
+     * retrieve the files before trying to load any specific pages.
```
Comment (johnsaigle) [inline modules/login/php/module.class.inc]: These lines should be updated to reflect the behaviour of _this_  `handle` function, not the general idea of `handle`

File: modules/login/php/module.class.inc [hunk 17]
```diff
@@ -3,35 +3,67 @@
...
+            \Module::factory('api');
+            $handler = new endpoints\Authentication();
+            break;
+        default:
+            return parent::handle($request);
+        }
+        $user = $request->getAttribute('user');
+        if (!$handler->_hasAccess($user)) {
+            return new \LORIS\Http\Response\JSON\Forbidden();
+        }
```
Comment (johnsaigle) [inline modules/login/php/module.class.inc]: I think You don't need this and instead this module should override the parent class's `isPublicModule()` function to be true. Every user should be able to access this module.

File: modules/login/php/module.class.inc [hunk 18]
```diff
@@ -3,35 +3,67 @@
...
 
     /**
-     * The login module must be accessible publicly, so that non-logged in
-     * users can login.
+     * Returns true if this module should be publicly available to people who
+     * are not logged in to LORIS.
      *
-     * @return true
+     * @return bool False. Subclasses must override this function in the case
+     *                  that they should be publicly available.
```
Comment (johnsaigle) [inline modules/login/php/module.class.inc]: This comment is inaccurate

File: modules/login/php/module.class.inc (hunk 17 above)
Comment (maltheism) [reply modules/login/php/module.class.inc]: @johnsaigle I'm going to do similar to this PR from the comments received from @driusan about the issue_tracker PR.

File: modules/login/php/endpoints/authentication.class.inc (hunk 1 above)
Comment (maltheism) [reply modules/login/php/endpoints/authentication.class.inc]: @johnsaigle I'm going to do similar to this PR from the comments received from @driusan about the issue_tracker PR.

File: modules/login/php/endpoints/authentication.class.inc (hunk 6 above)
Comment (maltheism) [reply modules/login/php/endpoints/authentication.class.inc]: Hi @johnsaigle can you make a user with factory from a username?

**Comment by cmadjar** [comment]
@kongtiaowang Could you help @maltheism fix Travis on this PR? Thank you!

**Comment by kongtiaowang** [comment]
@maltheism When I test this login module, If I input a wrong name or password, there is no error message to reminder user. ["Incorrect username or password"].
There is also missing welcome message.
If you could fix this. I will send you a commit to fix Travis.

File: modules/login/jsx/loginIndex.js [hunk 19]
```diff
@@ -0,0 +1,301 @@
...
+  }
+
+  /**
+   * Set the form data based on state values of child elements/components
+   *
+   * @param {string} formElement - name of the selected element
+   * @param {string} value - selected value for corresponding form element
+   */
+  setForm(formElement, value) {
+    const state = Object.assign({}, this.state);
```
Comment (HenriRabalais) [inline modules/login/jsx/loginIndex.js:96]: I'm pretty sure we should be getting in the habit of using `JSON.parse(JSON.stringify(this.state))` since `Object.assign()` only creates a shallow copy. This also applies to lines 

https://stackoverflow.com/questions/41847581/what-is-the-difference-between-object-assign-and-json-parsejson-stringifyobj/41848531

According to this benchmark test, it also appears to be the fastest:
https://www.measurethat.net/Benchmarks/Show/2471/2/objectassign-vs-json-stringparse

Comment (maltheism) [reply modules/login/jsx/loginIndex.js:96]: @HenriRabalais that's cool! How did you figure that out? Btw, I'm surprised there is nothing better than the `JSON.parse(JSON.stringify({..})` because I've seen or been told (whichever I cannot remember) that solution is a "hacky" way of making a new object. But that was years ago and maybe it's no longer considered hacky. I myself don't usually care if something is considered hacky by others if it works well! 😅

Comment (ridz1208) [reply modules/login/jsx/loginIndex.js:96]: @maltheism in 05 i still thought I was going to be an astronaut

Comment (HenriRabalais) [reply modules/login/jsx/loginIndex.js]: @maltheism I had been using it ever since Armin had recommended it as a proper way of cloning an object, and now every time I look into it when I doubt myself I see that it actually seems like the more efficient and robust solution!

**Comment by maltheism** [comment]
Thanks @cmadjar! That helped a lot..

File: modules/login/php/requestaccount.class.inc [hunk 20]
```diff
@@ -1,323 +0,0 @@
-<?php
```
Comment (ridz1208) [inline modules/login/php/requestaccount.class.inc:1]: @maltheism what's the reasoning behind getting rid of the individual classes/endpoints and combining them into one big authentication class with separate functions instead ?

These classes barely have anything in common, it seems more logical for separation of concerns to keep each one in it's own container.

Comment (maltheism) [reply modules/login/php/requestaccount.class.inc:1]: @ridz1208 Yah that can be done.

**Comment by ridz1208** [comment]
@maltheism can you make sure to use or improve upon the password element that has already been added to the codebase recently

**Comment by maltheism** [comment]
@ridz1208 yah I implemented that functionality already from looking at the PR awhile back.

**Comment by maltheism** [comment]
@ridz1208 I added it.

**Comment by ridz1208** [comment]
@maltheism conflicts

**Comment by maltheism** [comment]
@ridz1208 conflicts are fixed

**Comment by maltheism** [comment]
@ridz1208 needs review.

**Comment by laemtl** [comment]
Once rebased I can review.

**Comment by maltheism** [comment]
Hi @laemtl, I rebased the PR.

**Review by laemtl — CHANGES_REQUESTED** [review]
The following scenario does not allow me to connect (but does work if I switch on main):
- request an account
- login as admin, approve the account and generate password

Using the generated password return a 409 error status. The error parameter in the response is empty, so no feedback is displayed to the user, and nothing happens (no login, no password change prompt).

**Comment by laemtl** [comment]
Also, I noticed a minor problem, FormElement is missing a name attribute:

```
Warning: Failed prop type: The prop `name` is marked as required in `FormElement`, but its value is `null`.
FormElement@http://localhost:8091/js/components/Form.js:164:81
Login@http://localhost:8091/login/js/loginIndex.js:155:81
```

**Comment by maltheism** [comment]
Hi @laemtl, this PR is somewhat old and few things changed before this PR ever had a chance to get merged when it was working. The changes I didn't know about until today broke the expired password of my code from firing like it previously did. That should be fixed now and should be ready to be reviewed again. Thanks for catching it!

**Comment by laemtl** [comment]
@maltheism Everything seems to work really great!
One last thing: When I request a new password, I no longer get the screen to change the generated password.

**Comment by maltheism** [comment]
Hi @laemtl, PR is ready for review again and your comment helped. thanks!

**Comment by laemtl** [comment]
@maltheism I retested and I found another small problem (but we are getting closer!). If we reset the password, and pick something that does not pass the validation rules, the error is not displayed to the user and nothing happens (no submission). I used '1' in my tests.

**Review by laemtl — DISMISSED** [review]
I found another small problem (but we are getting closer!). If we reset the password, and pick something that does not pass the validation rules (ex: '1'), the error is not displayed to the user.

**Comment by maltheism** [comment]
Hi @laemtl @HenriRabalais @kongtiaowang can someone review again, pretty please? 🙂

File: modules/login/jsx/loginIndex.js [hunk 21]
```diff
@@ -0,0 +1,311 @@
...
+            />
+            <PasswordElement
+              name={'password'}
+              value={this.state.form.value.password}
+              onUserInput={this.setForm}
+              placeholder={'Password'}
+              class={'col-sm-12'}
+              required={true}
+              autoComplete={'current-password'}
+              isPassword={true}
```
Comment (HenriRabalais) [inline modules/login/jsx/loginIndex.js:219]: I don't see where this prop gets used.

**Review by HenriRabalais — COMMENTED** [review]
Just some clarification questions.

File: jsx/Form.js [hunk 22]
```diff
@@ -469,35 +469,6 @@ class SelectElement extends Component {
     this.handleChange = this.handleChange.bind(this);
   }
 
-  /**
```
[the remark this answers is not in the document]
Comment (maltheism) [reply jsx/Form.js:472]: Hi @HenriRabalais, it looks like a rebase mistake.

File: jsx/Form.js [hunk 23]
```diff
@@ -1122,7 +1091,7 @@ class TextboxElement extends Component {
     // and retain formatting. If label prop is not provided at all, the input
     // element will take up the whole row.
     let label = null;
-    let inputClass = 'col-sm-12';
+    let inputClass = this.props.class;
```
[the remark this answers is not in the document]
Comment (maltheism) [reply jsx/Form.js:1094]: Hi @HenriRabalais, the non-reactified login module has text inputs without the column classes or labels for part of the module. The bootstraps column class ruins the old login appearance. So this change made it so you could not have the column class and if you look where I used it.

File: jsx/Panel.js [hunk 24]
```diff
@@ -55,31 +55,38 @@ class Panel extends Component {
...
+
     // Add panel header, if title is set
     const panelHeading = this.props.title ? (
       <div
         className="panel-heading"
         onClick={this.toggleCollapsed}
         data-toggle="collapse"
         data-target={'#' + this.props.id}
-        style={{cursor: 'pointer', height: '3em', fontWeight: 'bold'}}
+        style={this.props.collapsing ?
```
[the remark this answers is not in the document]
Comment (maltheism) [reply jsx/Panel.js:71]: Hi @HenriRabalais, I like that idea and should I do that now in this PR or discuss at Loris meeting for new PR?

File: modules/login/jsx/loginIndex.js (hunk 21 above)
Comment (maltheism) [reply modules/login/jsx/loginIndex.js:219]: @HenriRabalais thanks. I forgot to remove that prop when I changed the TextElement component to the PasswordElement component.

Comment (maltheism) [reply modules/login/jsx/loginIndex.js:219]: @laemtl do you know if eslint has a flag for giving props that don't exist?

File: jsx/Panel.js (hunk 24 above)
[the remark this answers is not in the document]
Comment (HenriRabalais) [reply jsx/Panel.js:71]: New PR for sure! I think this one is already packed with a lot of stuff.

File: jsx/Form.js (hunk 23 above)
[the remark this answers is not in the document]
Comment (HenriRabalais) [reply jsx/Form.js:1094]: Hmm, yeah. Makes sense. I think we'll just need to do a significant refactor of Form.js where there's different wrappers that meet the various styling needs, either through boolean values or through different parent component name. So something like:
```
<Form>
  <Text/>
  <Text/>
  <Select/>
  ...
</Form>
```
would just be our standard. But then we could have
```
<Form placeholder={true}>
  <Text/>
  <Text/>
  <Select/>
  ...
</Form>
```
where in both cases, the `Form` would take care of wrapping all the children in the proper styling.

That way, you could also have stuff like:
```
<InlineForm>
  <Text/>
  <Text/>
  <Select/>
  ...
</InlineForm>
```
where there's multiple fields on the same line.

File: modules/login/jsx/loginIndex.js (hunk 21 above)
Comment (laemtl) [reply modules/login/jsx/loginIndex.js]: Here is the list of available rules for react: https://github.com/yannickcr/eslint-plugin-react#list-of-supported-rules

Comment (maltheism) [reply modules/login/jsx/loginIndex.js]: `forbidExtraProps`

**Review by laemtl — DISMISSED** [review]
The following scenario needs adjustements:
- click Forgot your password?
- Use the temporary password to login
- Entering mismatched password allows to save (first one is saved).

**Comment by maltheism** [comment]
@laemtl fixed

**Comment by driusan** [comment]
@maltheism @laemtl any update on this?

**Comment by maltheism** [comment]
Hi @driusan, the PR is needing a review and I think @laemtl will when she has time.

**Review by laemtl — CHANGES_REQUESTED** [review]
Items 4, 6, 7, 8, 9 and 10 from this checklist are failing: https://github.com/aces/Loris/blob/main/modules/login/test/Reset_Password_Test_Plan.md

I wasn't able to change my password, even with one that passes validation.

**Review by laemtl — APPROVED** [review]

**Review by laemtl — CHANGES_REQUESTED** [review]
Item 8. for the test plan is failing: 
`Try entering the user's username as the new password. You should get an error.`

**Review by laemtl — APPROVED** [review]
